### ISSUE TITLE
improve the camera component

### ISSUE THREAD

**Issue by halatmit** [mit-cml/appinventor-sources#338]
Access use the Camera API to improve the Camera Compnent.   Espcially the ability to control fron-tfacing vs back-facing, and image orientation.

See

http://developer.android.com/guide/topics/media/camera.html

**Comment by halatmit** [comment]
also see pull request #335

**Comment by marksherman** [comment]
This is absolutely the right answer.

**Comment by josmas** [comment]
is this taken care of by #363 ?

**Comment by halatmit** [comment]
@josmas 
I don't think that #363 covers this, but I'm not sure.

**Comment by josmas** [comment]
#363 is re-writing the component based on an [old design from GSoC12] (https://github.com/josmas/OurBricks_AppInventor/wiki/Live-Feed-Design) that uses a surface view to do the live feed, instead of the default camera, so this should be scoped into that work.
@thequixotic are you going to work on this over the summer?

**Comment by sjais789** [comment]
Camera API 2 supports to check whether your device is using front camera or rear camera using camera   **getCameraIdList()** method and  **getCameraCharacteristics()** to add properties like use front_camera or not.
But camera 2 API supports on  Android API 21+ (Android 5 - Lollipop).

**Comment by thequixotic** [comment]
We support App Inventor from API Level 4. We try to be consistent with features so that they are evident on most devices. I don't know if its nice to include it as it might disappoint most users (who use devices below API 21).

**Comment by sjais789** [comment]
I agree with it but if can we try two code paths for API (>21+) and API (<21+) to solve our front camera issues. @josmas @thequixotic .

![capture](https://cloud.githubusercontent.com/assets/12415740/24251042/d47141d6-0ffe-11e7-904f-209d70446c4e.PNG)

`)

**Comment by marksherman** [comment]
What would the semantics of the blocks language be to handle that with
maximal compatibility?

On Thu, Mar 23, 2017 at 9:58 AM Shubham <notifications@github.com> wrote:

I agree with it but if can we try two code paths for API (>21+) and API
(<21+) to solve our front camera issues. @josmas
<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_josmas&d=DgMFaQ&c=lqHimbpwJeF7VTDNof4ddl8H-RbXeAdbMI2MFE1TXqA&r=LSK4ZJvCmGVZ1u3kVxFdcOB46IO-bbzvhavuoIxMLKQ&m=Zkd_LdnU1_ebnli00whGa5KXYrk6DYgueCvTdr3ZG1s&s=dETF0NofzbA_upu_bGEmhPZzxpHkeDchNiwDmwfZ_bA&e=>
@thequixotic
<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_thequixotic&d=DgMFaQ&c=lqHimbpwJeF7VTDNof4ddl8H-RbXeAdbMI2MFE1TXqA&r=LSK4ZJvCmGVZ1u3kVxFdcOB46IO-bbzvhavuoIxMLKQ&m=Zkd_LdnU1_ebnli00whGa5KXYrk6DYgueCvTdr3ZG1s&s=LuQmC-snfjMy9xz28HUSmPqfFDnwsEWXq9pGnT7TM8c&e=>
.

[image: capture]
<https://urldefense.proofpoint.com/v2/url?u=https-3A__cloud.githubusercontent.com_assets_12415740_24251042_d47141d6-2D0ffe-2D11e7-2D904f-2D209d70446c4e.PNG&d=DgMFaQ&c=lqHimbpwJeF7VTDNof4ddl8H-RbXeAdbMI2MFE1TXqA&r=LSK4ZJvCmGVZ1u3kVxFdcOB46IO-bbzvhavuoIxMLKQ&m=Zkd_LdnU1_ebnli00whGa5KXYrk6DYgueCvTdr3ZG1s&s=1zipEkUW_mfHnYbv6YFzslwRSSeHc3qgytH3vOHtJIY&e=>

`)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_mit-2Dcml_appinventor-2Dsources_issues_338-23issuecomment-2D288727314&d=DgMFaQ&c=lqHimbpwJeF7VTDNof4ddl8H-RbXeAdbMI2MFE1TXqA&r=LSK4ZJvCmGVZ1u3kVxFdcOB46IO-bbzvhavuoIxMLKQ&m=Zkd_LdnU1_ebnli00whGa5KXYrk6DYgueCvTdr3ZG1s&s=yzHb7oQpDKvzEHHua_mWTEvYGE_8p0BNj354pO51awA&e=>,
or mute the thread
<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AB8NzT71eOhXeknRd91KXhA9MXKgpok-5Fks5ronqagaJpZM4D2Jih&d=DgMFaQ&c=lqHimbpwJeF7VTDNof4ddl8H-RbXeAdbMI2MFE1TXqA&r=LSK4ZJvCmGVZ1u3kVxFdcOB46IO-bbzvhavuoIxMLKQ&m=Zkd_LdnU1_ebnli00whGa5KXYrk6DYgueCvTdr3ZG1s&s=YFFOxk0Vy1iD0pRM4M3CcMC3-ON4qyu8w5F7rqGLFPI&e=>
.

-- 
Mark Sherman
UMass Lowell Computer Science

**Comment by Ang-YC** [comment]
In Block Language, the basic functionalities should include:

```
when [Camera].Error (message) do
when [Camera].AfterPicture (image) do
when [Camera].Switched do

call [Camera].Open
call [Camera].Close
call [Camera].TakePicture
call [Camera].SwitchCamera

set [Camera].Front to (true/false)

get [Camera].IsOpen (true/false)
get [Camera].Front (true/false)
get [Camera].HaveFront (true/false)
get [Camera]
```

When `[Camera].Front` is set to true, it will start the camera using Front Camera, user can also check this value to know if current Camera is Front or Back.

And also other methods inherited from `AndroidViewComponent`
```
[Camera].Width, [Camera].Height, [Camera].Visible
```

This will be implemented using custom SurfaceView with no controls, user can use other components to call the method to take picture.

**Comment by marksherman** [comment]
That sounds good. What do those blocks do when running on a device with a
lower API level?

> --
Mark Sherman
UMass Lowell Computer Science

**Comment by Ang-YC** [comment]
I have submitted the code few days ago for GSoC 17 (I hope you are one of them that can read the application) which also includes the link to my modified App Inventor server with Embedded Camera feature. I am not planning to post the link here to prevent server abuse anyway. (Can I send you an email of the link so you can try it on?)

It's tested on Android 2.3, 4.1, 6.0 (real device), Android 2.0, 2.1 (emulator), I think it screws up on Android 1.6 (Not sure what's the problem, logcat shows `java.lang.VerifyError: Verifier rejected class`)
Compiled APK: https://drive.google.com/file/d/0B-XsmXCo7K41anNOdkRvVThYcmc/view?usp=sharing

**Comment by ewpatton** [comment]
I feel like a surface based view rather than launching the camera app would get tricky with the existing non-visible component. We may want to propose a new component with visual representation in the designer (similar to PIC, which does do the live camera feed). A design doc in a new issue would be welcome but it is not something the MIT team should take up at this time.
